### PR TITLE
Fix zero-jvps with shot vectors

### LINKED ISSUE #6220: [BUG] Zero-valued JVPs of gradient-free tapes with shot vectors have the wrong shape
### Expected behavior

Found by @astralcai 

Gradient-free tapes with shot vectors can be differentiated, giving a zero with the right output shape.

### Actual behavior

The output shape is off.

### Additional information

_No response_

### Source code

```shell
import jax
import jax.numpy as jnp
import pennylane as qml

dev = qml.device("default.qubit", wires=1)

def cost(theta):
    tape1 = qml.tape.QuantumScript([qml.RX(0.5, wires=0)], [qml.expval(qml.Z(0))], shots=[20, 20])
    tape2 = qml.tape.QuantumScript([qml.RY(theta, wires=0)], [qml.expval(qml.Z(0))], shots=[20, 20])
    res1, res2 = qml.execute([tape1, tape2], dev, gradient_fn=qml.gradients.param_shift)
    return res1, res2

params = jnp.array(0.5)
jax.jacobian(cost)(params)
```

### Tracebacks

```shell
TypeError: Custom JVP rule _execute_and_compute_jvp for function _execute_wrapper must produce primal and tangent outputs with equal container (pytree) structures, but got PyTreeDef(((*, *), (*, *))) and PyTreeDef((*, (*, *))) respectively.
```

### System information

```shell
pl dev
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
If a tape has no trainable parameters, generic zero-valued JVPs are created for it in JVP calculations.
However, these generic calculations are not taking shot vectors correctly into account, also because they do not use `MeasurementProcess.shape` correctly.

**Description of the Change:**
Change the generic zero-valued JVPs so they are compatible with shot vectors.

**Benefits:**
One bug less.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Fixes #6220 

[sc-73033]

### CODE REVIEW

**Review by astralcai — APPROVED** [review]
LGTM, thank you for fixing this!

File: pennylane/gradients/jvp.py [hunk 1]
```diff
@@ -295,11 +295,16 @@ def jvp(tape, tangent, gradient_fn, gradient_kwargs=None):
...
-            res = tuple(np.zeros(mp.shape(None, tape.shots)) for mp in tape.measurements)
+        def zero_jvp_for_single_shots(s):
+            res = tuple(np.zeros(mp.shape(shots=s)) for mp in tape.measurements)
             return res[0] if len(tape.measurements) == 1 else res
 
-        return tuple(), zero_vjp
+        def zero_jvp(_):
+            if tape.shots.has_partitioned_shots:
+                return tuple(zero_jvp_for_single_shots(s) for s in tape.shots)
+            return zero_jvp_for_single_shots(tape.shots)
```
Comment (albi3ro) [inline pennylane/gradients/jvp.py:305]:
```suggestion
            return zero_jvp_for_single_shots(tape.shots.total_shots)
```

File: pennylane/gradients/jvp.py [hunk 2]
```diff
@@ -295,11 +295,16 @@ def jvp(tape, tangent, gradient_fn, gradient_kwargs=None):
     if len(tape.trainable_params) == 0:
         # The tape has no trainable parameters; the JVP
         # is simply none.
-        def zero_vjp(_):
-            res = tuple(np.zeros(mp.shape(None, tape.shots)) for mp in tape.measurements)
+        def zero_jvp_for_single_shots(s):
+            res = tuple(np.zeros(mp.shape(shots=s)) for mp in tape.measurements)
```
Comment (albi3ro) [inline pennylane/gradients/jvp.py:299]: Do we also need to add in the numeric type?
```suggestion
            res = tuple(np.zeros(mp.shape(shots=s), dtype=mp.numeric_type) for mp in tape.measurements)
```

File: pennylane/workflow/jacobian_products.py [hunk 3]
```diff
@@ -46,6 +46,17 @@ def _compute_vjps(jacs, dys, tapes):
...
 
+def _zero_jvp_single_shots(shots, tape):
+    jvp = tuple(np.zeros(mp.shape(shots=shots), dtype=mp.numeric_type) for mp in tape.measurements)
+    return jvp[0] if len(tape.measurements) == 1 else jvp
+
+
+def _zero_jvp(tape):
+    if tape.shots.has_partitioned_shots:
+        return tuple(_zero_jvp_single_shots(s, tape) for s in tape.shots)
+    return _zero_jvp_single_shots(tape.shots, tape)
```
Comment (albi3ro) [inline pennylane/workflow/jacobian_products.py:57]:
```suggestion
    return _zero_jvp_single_shots(tape.shots.total_shots, tape)
```

File: pennylane/gradients/jvp.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/jvp.py:299]: Ah good point!

**Review by albi3ro — APPROVED** [review]

**Pull request merged**